core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java

```diff
@@ -410,9 +411,16 @@ private Executor resolveAsyncFutureCompletionExecutor(SdkClientConfiguration con
      * Finalize the internal SDK scheduled executor service that is used for scheduling tasks such
      * as async retry attempts and timeout task.
      */
-    private ScheduledExecutorService resolveScheduledExecutorService() {
-        return Executors.newScheduledThreadPool(5, new ThreadFactoryBuilder()
-            .threadNamePrefix("sdk-ScheduledExecutor").build());
+    private ScheduledExecutorService resolveScheduledExecutorService(SdkClientConfiguration config) {
+        Supplier<ScheduledExecutorService> defaultScheduledExecutor = () -> {
+            ScheduledExecutorService executor = Executors.newScheduledThreadPool(5, new ThreadFactoryBuilder()
+                .threadNamePrefix("sdk-ScheduledExecutor").build());
+
+            return executor;
```

Comment:
We need to make sure the custom ScheduledExecutorService doesn't get closed when the client is closed. We can create a `NonManagedScheduledExecutorService` wrapper similar to [`NonManagedSdkHttpClient`](https://github.com/aws/aws-sdk-java-v2/blob/master/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/builder/SdkDefaultClientBuilder.java#L526)/`NonManagedSdkAsyncHttpClient`. Currently, the executor is closed [here](https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/AttributeMap.java#L94) when the client is closed. We can override `shutdown()` to do nothing:

```
    /**
     * Wrapper around {@link ScheduledExecutorService} to prevent it from being closed. Used when the customer provides
     * a custom scheduled executor service in which case they are responsible for the lifecycle of it.
     */
    @SdkTestInternalApi
    public static final class NonManagedScheduledExecutorService implements ScheduledExecutorService {

        private final ScheduledExecutorService delegate;

        NonManagedScheduledExecutorService(ScheduledExecutorService delegate) {
            this.delegate = paramNotNull(delegate, "ScheduledExecutorService");
        }

        public ScheduledExecutorService scheduledExecutorService() {
            return delegate;
        }

        @Override
        public void shutdown() {
            // Do nothing, this executor service is managed by the customer.
        }

        @Override
        public List<Runnable> shutdownNow() {
            return new ArrayList<>();
        }

        @Override
        public boolean isShutdown() {
            return delegate.isShutdown();
        }

        @Override
        public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
            return delegate.awaitTermination(timeout, unit);
        }

       // override remaining methods with delegate.method()...
```

```
    private ScheduledExecutorService resolveScheduledExecutorService(SdkClientConfiguration config) {
        ScheduledExecutorService customScheduledExecutorService = config.option(SCHEDULED_EXECUTOR_SERVICE);
        if (customScheduledExecutorService != null) {
            return new NonManagedScheduledExecutorService(customScheduledExecutorService);
        }

        return Executors.newScheduledThreadPool(5, new ThreadFactoryBuilder()
            .threadNamePrefix("sdk-ScheduledExecutor").build());
    }
```

Let's also add a test for this in [`ResourceManagementTest`](https://github.com/aws/aws-sdk-java-v2/blob/master/test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/ResourceManagementTest.java):

```
    @Test
    public void scheduledExecutorFromBuilderNotShutdown() {
        ScheduledExecutorService scheduledExecutorService = mock(ScheduledExecutorService.class);

        asyncClientBuilder().overrideConfiguration(c -> c.scheduledExecutorService(scheduledExecutorService)).build().close();

        verify(scheduledExecutorService, never()).shutdown();
        verify(scheduledExecutorService, never()).shutdownNow();
    }
```

Reply:
Would you prefer to create a utility class like https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/ExecutorUtils.java ? `SdkDefaultClientBuilder` is already big and there is a lot of methods to implements in `ScheduledExecutorService`.

Also, the future completion executor is wrapped in the configuration builder class rather in the client builder, should we keep consistency and wrap the `scheduledExecutorService` in the `ClientOverrideConfiguration:Builder` class ?

Reply:
Yeah that sounds good, maybe `ScheduledExecutorUtils`? 

Makes sense, let's put it in `ClientOverrideConfiguration:Builder` instead.

Reply:
Make sense, TBH, there is a lot of piece of code everywhere that create thread pool and its not really easy to follow and find out where to override it when possible.
I'll keep it there and maybe raise an issue to refactor the rest ?

Reply:
Agreed, yup sounds good to me